Yoast/javascript
# P2 216 Create warning block (to show when Schema blocks are removed)

## Summary


This PR can be summarized in the following changelog entry:

* Adds a warning block, to be shown when required or recommended Schema blocks are removed in the editor.

## Relevant technical choices:

* The warning block is registered in the `initialize` method in `index.ts`.
* The warning block has only been created, but as of yet there is no implementation for it. It will be implemented in https://yoast.atlassian.net/browse/P2-217.
* We added keys to all the elements in `edit.ts` to prevent a console warning ("each child in a list should have a unique "key" prop").

## Test instructions

This PR can be tested by following these steps:

* Make sure the Schema blocks feature flag is on, and the monorepo is linked to the Free plugin.
* You can test the looks of the warning block by replacing line 16-17 and before in `edit.ts` by the following code:

```js
	const attributes = {
		removedBlock: "core/paragraph",
		removedAttributes: {
			content: "The content of the paragraph.",
		},
		warningText: "You just removed a paragraph, do you really want to remove it?",
		isRequired: true,
	};

	const { removedBlock, removedAttributes, warningText, isRequired } = attributes;
```
* Comment out this line in `schema-blocks/src/blocks/warning-block/configuration.ts`:
  ```
  parent: [],
  ```
  * This makes sure that the Warning block can be added as a separate block in the editor.
* In your block editor, add a Warning block.
* Its styling should match the designs for [required](https://www.sketch.com/s/4c91f078-4831-45b9-8e95-ff171700ed27/a/4ywEYJ) and [recommended](https://www.sketch.com/s/4c91f078-4831-45b9-8e95-ff171700ed27/a/0w3zz0#Inspector) blocks.
* If you set `isRequired` in the code you pasted in the `edit` method to `true`, you should see the red warning for a required block. If you set `isRequired` to `false`, you should see the yellow warning for a recommended block.
* After each change you've made to the code, run `yarn build` in the `schema-blocks` directory in `javascript`, and `grunt build:dev` in `wordpress-seo`.
* In the Gutenberg editor, test that clicking 'yes' in the warning block removes the warning block.
* In the editor, test that clicking 'no' restores the pre-existing block.
  * Since there is no pre-existing block, we simulated one using the warning block attributes.
  * After clicking 'no' the warning block should be replaced by a paragraph with "The content of the paragraph." as its content. 	

### Test instructions for QA when the code is in the RC



QA can test this PR by following these steps:

* Add a Yoast Schema block (recipe or job posting).
* Add a block that is required for that Schema block.
* Remove the required block. 
* You should see a warning block that follows this design: https://www.sketch.com/s/4c91f078-4831-45b9-8e95-ff171700ed27/a/4ywEYJ
* Add a block that is recommended for that Schema block.
* Remove the recommended block. 
* You should see a warning block that follows this design: https://www.sketch.com/s/4c91f078-4831-45b9-8e95-ff171700ed27/a/0w3zz0
* Click "Yes" to restore the block, and see it is restored. You should see the corresponding Schema output on the frontend.
* Click "No", and see the block isn't restored.

## Impact check

* This PR affects the following parts of the plugin, which may require extra testing:
  * The structured data blocks.

## UI changes

## Quality assurance


Fixes https://yoast.atlassian.net/browse/P2-216
